nsqd/gossip.go

```diff
@@ -0,0 +1,402 @@
+		Topic:   topicName,
+		Channel: channelName,
+		Rnd:     time.Now().UnixNano(),
+	}
+	payload, err := json.Marshal(&gev)
+	if err != nil {
+		return err
+	}
+	return n.serf.UserEvent(evName, payload, false)
+}
+
+func (n *NSQD) gossipLoop() {
+	var evName string
+	var topicName string
+	var channelName string
+
+	if n.serf.EncryptionEnabled() {
+		err := n.rotateGossipKey()
+		n.logf("FATAL: could not rotate gossip key - %s", err)
+		n.Exit()
+		return
+	}
+
+	regossipTicker := time.NewTicker(n.opts.GossipRegossipInterval)
+
+	if len(n.opts.GossipSeedAddresses) > 0 {
+		for {
+			num, err := n.serf.Join(n.opts.GossipSeedAddresses, false)
+			if err != nil {
+				n.logf("ERROR: failed to join serf - %s", err)
+				select {
+				case <-time.After(15 * time.Second):
+					// keep trying
+				case <-n.exitChan:
+					goto exit
+				}
+			}
+			if num > 0 {
+				n.logf("SERF: joined %d nodes", num)
+				break
+			}
+		}
+	}
+
+	for {
+		select {
+		case <-regossipTicker.C:
+			n.logf("SERF: re-gossiping")
+			stats := n.GetStats()
+			for _, topicStat := range stats {
+				if len(topicStat.Channels) == 0 {
+					// if there are no channels we just send a topic exists event
+					err := n.gossip("topic=", topicStat.TopicName, "")
+					if err != nil {
+						n.logf("ERROR: failed to send Serf user event - %s", err)
+					}
+					continue
+				}
+				// otherwise only send a channel, implying the existence of the topic
+				for _, channelStat := range topicStat.Channels {
+					err := n.gossip("channel=", topicStat.TopicName, channelStat.ChannelName)
+					if err != nil {
+						n.logf("ERROR: failed to send Serf user event - %s", err)
+					}
+				}
+			}
+		case v := <-n.gossipChan:
+			switch v.(type) {
+			case *Channel:
+				channel := v.(*Channel)
+				topicName = channel.topicName
+				channelName = channel.name
+				if !channel.Exiting() {
+					evName = "channel+"
+				} else {
+					evName = "channel-"
+				}
+			case *Topic:
+				topic := v.(*Topic)
+				topicName = topic.name
+				channelName = ""
+				if !topic.Exiting() {
+					evName = "topic+"
+				} else {
+					evName = "topic-"
+				}
+			}
+			err := n.gossip(evName, topicName, channelName)
+			if err != nil {
+				n.logf("ERROR: failed to send Serf user event - %s", err)
+			}
+		case <-n.exitChan:
+			goto exit
+		}
+	}
+
+exit:
+	regossipTicker.Stop()
+	n.logf("GOSSIP: exiting")
+}
+
+func (n *NSQD) initialGossipKey() []byte {
+	var key []byte
+	if n.tlsConfig != nil && len(n.tlsConfig.Certificates) > 0 {
+		key = n.tlsConfig.Certificates[0].Leaf.Signature
+	}
+	return key
+}
+
+func (n *NSQD) rotateGossipKey() error {
+	key := make([]byte, 32)
+	_, err := rand.Reader.Read(key)
+	strKey := base64.StdEncoding.EncodeToString(key)
+	_, err = n.serf.KeyManager().InstallKey(strKey)
+	if err != nil {
+		return err
+	}
+	_, err = n.serf.KeyManager().UseKey(strKey)
+	if err != nil {
+		return err
+	}
+	_, err = n.serf.KeyManager().RemoveKey(string(n.initialGossipKey()))
```

Comment:
does this always need to be the *last* key - it seems like after the first loop, it won't remove anything, right?